substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,581 @@
+            if (!(inlineInvokePlugin instanceof NativeImageInlineDuringParsingPlugin)) {
+                result.getPlugins().appendInlineInvokePlugin(inlineInvokePlugin);
+            }
+        }
+
+        return result;
+    }
+
+    @SuppressWarnings("try")
+    InvocationResult analyzeMethod(CallSite callSite, AnalysisMethod method, ValueNode[] args) {
+        if (!method.hasBytecodes()) {
+            /* Native method. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        } else if (providers.getGraphBuilderPlugins().getInvocationPlugins().lookupInvocation(method) != null) {
+            /* Method has an invocation plugin that we must not miss. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        } else if (method.isSynchronized()) {
+            /*
+             * Synchronization operations will always bring us above the node limit, so no point in
+             * starting an analysis.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+
+        MethodNodeTrackingAndInline methodState = new MethodNodeTrackingAndInline(new InvocationResultInline(callSite, method));
+
+        GraphBuilderConfiguration graphBuilderConfig = prototypeGraphBuilderConfig.copy();
+        graphBuilderConfig.getPlugins().appendInlineInvokePlugin(methodState);
+        graphBuilderConfig.getPlugins().appendParameterPlugin(new TrivialMethodDetectorParameterPlugin(args));
+
+        StructuredGraph graph = new StructuredGraph.Builder(options, debug).method(method).build();
+
+        try (DebugContext.Scope ignored = debug.scope("InlineDuringParsingAnalysis", graph, method, this)) {
+
+            TrivialMethodDetectorGraphBuilderPhase builderPhase = new TrivialMethodDetectorGraphBuilderPhase(bb, providers, graphBuilderConfig, OptimisticOptimizations.NONE, null,
+                            providers.getWordTypes());
+
+            try (NodeEventScope ignored1 = graph.trackNodeEvents(methodState)) {
+                builderPhase.apply(graph);
+            }
+
+            debug.dump(DebugContext.VERBOSE_LEVEL, graph, "InlineDuringParsingAnalysis successful");
+            return methodState.result;
+        } catch (Throwable ex) {
+            String description = "InlineDuringParsingAnalysis failed with " + ex.toString();
+            debug.dump(DebugContext.VERBOSE_LEVEL, graph, description);
+            /*
+             * Whatever happens during the analysis is non-fatal because we can just not inline that
+             * invocation.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+    }
+
+    /**
+     * We collect information during graph construction to filter non-trivial methods and inline
+     * trivial invokes. The result is used in {@link #analyzeMethod}.
+     */
+    class MethodNodeTrackingAndInline extends NodeEventListener implements InlineInvokePlugin {
```

Comment:
I think this can be split into two classes one that `extends NodeEventListener` and one that `implements InlineInvokePlugin`, there is no shared state between the two so I think it is cleaner if they are separated, just like you do with the `TrivialMethodDetectorParameterPlugin`.

Reply:
Done, thanks for the suggestion.